### DISCUSSION TITLE
Enumerate available peripherals via HTTP commands

### DISCUSSION THREAD

**Discussion by kryslt** [arendst/Tasmota#20164] [Q&A] [answered]
Hello, is it possible to enumerate available peripherals via HTTP commands?
Request. like:
`http://<ip>/cm?cmnd=EnumPer`
Response like:
`{"Relay":2,"Powermeter":1,"Lightsensor":1,"Tempsensor":4}`

Then get individual peripheral type:
Request: 
`http://<ip>/cm?cmnd=PerType%20Relay1`
Response:
`{"Relay1":"Electromagnetic"}`

Request: 
`http://<ip>/cm?cmnd=PerType%20Relay2`
Response:
`{"Relay2":"Triac"}`

Request: 
`http://<ip>/cm?cmnd=PerType%20Powermeter1`
Response:
`{"Powermeter1":"ADE7753ARSZ"}`

**Comment by barbudor** [comment]
What do you mean by "enumerate peripherals"?

If you know how to do it by command in the console, then you can also do it by Http or mqtt

**Reply by kryslt** [reply]
Sorry for being not specific. I have edited the question.

**Reply by sfromis** [reply]
Tasmota cannot know what you have wired to a Relay pin, it could also be a led, supplying power to a sensor, or whatever.
Using `Gpio all`, you get what pins you have configured for specific behaviors, including Relay pins, or energy monitoring chips.
Using `Status 10`, you get sensors and values, including I2C sensors not configured as gpio component types.
If you make sure to disable unused drivers, `I2CDriver` will get you what I2C drivers are enabled.

**Reply by kryslt** [reply]
Tasmota surely not but the device manufacturer who uses it. There should be some table that informs which peripherals are connected. Gpio all is something different, I'm not interested in pins but attached peripherals.
EDIT: Status 10 message is also not a solution. It returns status of only some of the attached peripherals, not short listing of all of them.

**Reply by sfromis** [reply]
When the device is flashed with Tasmota, any information the hardware manufacturer may (or not) have had in factory firmware is no longer present, thus no way of knowing what's physically connected to a pin. `Status 10` works great for sensors, which obviously was not meant as a full answer to what you'd like to have. Likewise gpio is very much relevant for many types of devices Tasmota where can know what it is. That does not include whether it is a physical relay and so on, only what's known by Tasmota.

In short, you have no feature for how the hardware is built, and it would not even be possible for Tasmota to have such a feature. What you *can* have is information about how Tasmota is configured for the hardware, and in some cases (mostly I2C), what Tasmota could recognize on a bus. There are specific ways of gathering info on many of the features, but not the way you imagined.

**Reply by kryslt** [reply]
Yep, this is exactly what I miss. I want to know how Tasmota is configured for the hardware in a simple short listing as I proposed.

**Reply by sfromis** [reply]
I already told you of ways to see what's configured, but you did not like the `gpio all` command to see what's configured. And no, you are not going the get a "simple short list" matching your wish.

**Reply by kryslt** [reply]
Yes, i know already. This would require a completely new feature to be implemented. Thank you anyway for your time and provided information.

**Comment by kryslt** [comment, marked as answer]
From the above discussion I reply to myself: There is no such feature.
